module/zfs/dmu_recv.c

```diff
@@ -1374,29 +1438,147 @@ receive_object(struct receive_writer_arg *rwa, struct drr_object *drro,
+		 * using the ZPL's generation number to make the above
+		 * distinction.
+		 */
+		if (drro->drr_blksz != doi.doi_data_block_size) {
+			if (rwa->raw) {
+				/*
+				 * RAW streams always have large blocks, so
+				 * we are sure that the data is not needed
+				 * due to changing --large-block to be on.
+				 * Which is fortunate since the bonus buffer
+				 * (which contains the ZPL generation) is
+				 * encrypted, and the key might not be
+				 * loaded.
+				 */
+				do_free_range = B_TRUE;
+			} else if (rwa->full) {
+				/*
+				 * This is a full send stream, so it always
+				 * replaces what we have.  Even if the
+				 * generation numbers happen to match, this
+				 * can not actually be the same logical file.
+				 * This is relevant when receiving a full
+				 * send as a clone.
+				 */
+				do_free_range = B_TRUE;
+			} else if (drro->drr_type !=
+			    DMU_OT_PLAIN_FILE_CONTENTS ||
+			    doi.doi_type != DMU_OT_PLAIN_FILE_CONTENTS) {
+				/*
+				 * PLAIN_FILE_CONTENTS are the only type of
+				 * objects that have ever been stored with
+				 * large blocks, so we don't need the special
+				 * logic below.  ZAP blocks can shrink (when
+				 * there's only one block), so we don't want
+				 * to hit the error below about block size
+				 * only increasing.
+				 */
+				do_free_range = B_TRUE;
+			} else if (doi.doi_max_offset <=
+			    doi.doi_data_block_size) {
+				/*
+				 * There is only one block.  We can free it,
+				 * because its contents will be replaced by a
+				 * WRITE record.  This can not be the no-L ->
+				 * -L case, because the no-L case would have
+				 * resulted in multiple blocks.  If we
+				 * supported -L -> no-L, it would not be safe
+				 * to free the file's contents.  Fortunately,
+				 * that is not allowed (see
+				 * recv_check_large_blocks()).
+				 */
+				do_free_range = B_TRUE;
+			} else {
+				boolean_t is_same_gen;
+				err = receive_object_is_same_generation(rwa->os,
+				    drro->drr_object, doi.doi_bonus_type,
+				    drro->drr_bonustype, data, &is_same_gen);
+				if (err != 0)
+					return (SET_ERROR(EINVAL));
+
+				if (is_same_gen) {
+					/*
+					 * This is the same logical file, and
+					 * the block size must be increasing.
+					 * It could only decrease if
+					 * --large-block was changed to be
+					 * off, which is checked in
+					 * recv_check_large_blocks().
+					 */
+					if (drro->drr_blksz <=
+					    doi.doi_data_block_size)
+						return (SET_ERROR(EINVAL));
+					/*
+					 * We keep the existing blocksize and
+					 * contents.
+					 */
+					new_blksz =
+					    doi.doi_data_block_size;
+				} else {
+					do_free_range = B_TRUE;
+				}
+			}
+		}
+
+		/* nblkptr can only decrease if the object was reallocated */
+		if (nblkptr < doi.doi_nblkptr)
+			do_free_range = B_TRUE;
+
+		/* number of slots can only change on reallocation */
+		if (dn_slots != doi.doi_dnodesize >> DNODE_SHIFT)
+			do_free_range = B_TRUE;
+
+		/*
```

Comment:
Should we maybe refactor some of this into a separate function? receive_object is getting pretty large.

Reply:
For sure.  I was able to refactor some of the code for handling an existing object out into its own function.  Take a look and let me know what you think.